asm-to-pil/src/vm_to_constrained.rs

```diff
@@ -454,6 +456,86 @@ impl<T: FieldElement> ASMPILConverter<T> {
         res
     }
 
+    fn handle_instr_declaration_params(&mut self, m: &mut Machine<T>) {
+        for l in &mut m.links {
+            let lhs = &l.params;
+            let rhs = &mut l.to.params;
+            // lhs params must all be assignment registers when mapping instruction to operation
+            lhs.inputs.iter().chain(lhs.outputs.iter()).for_each(|p| {
+                assert!(p.ty.is_none());
+                assert!(self
+                    .registers
+                    .get(&p.name)
+                    .is_some_and(|r| r.ty.is_assignment()));
+            });
+
+            // if rhs is not empty, check it's valid
+            if !rhs.is_empty() {
+                // rhs params must either be assignment registers declared in the lhs or write registers
+                rhs
+                .inputs
+                .iter_mut()
+                .chain(rhs.outputs.iter_mut())
+                .for_each(|p| {
+                    assert!(p.ty.is_none());
```

Comment:
why is this true?

Reply:
This `Param.ty` here comes from parsing text of the form: `arg1: type1, arg2: type2 -> arg3: type3`.
We're currently using this same `Param` type to parse every kind of parameter declaration (instructions and functions etc), even when the `:type` part is not allowed, as is the case here, thus the assertion.
For example, you could try to do `instr a = somevm.func a: label`, which currently gets parsed, but is invalid.

Reply:
ah I see